io.openems.edge.io.shelly/src/io/openems/edge/io/shelly/shellyplugs/IoShellyPlugsImpl.java

```diff
@@ -0,0 +1,270 @@
+import io.openems.edge.timedata.api.utils.CalculateEnergyFromPower;
+
+@Designate(ocd = Config.class, factory = true)
+@Component(//
+		name = "IO.Shelly.PlugS", //
+		immediate = true, //
+		configurationPolicy = ConfigurationPolicy.REQUIRE//
+)
+@EventTopics({ //
+		EdgeEventConstants.TOPIC_CYCLE_EXECUTE_WRITE, //
+		EdgeEventConstants.TOPIC_CYCLE_AFTER_PROCESS_IMAGE //
+})
+public class IoShellyPlugsImpl extends AbstractOpenemsComponent implements IoShellyPlugs, DigitalOutput,
+		SinglePhaseMeter, ElectricityMeter, OpenemsComponent, TimedataProvider, EventHandler {
+
+	private final CalculateEnergyFromPower calculateProductionEnergy = new CalculateEnergyFromPower(this,
+			ElectricityMeter.ChannelId.ACTIVE_PRODUCTION_ENERGY);
+	private final CalculateEnergyFromPower calculateConsumptionEnergy = new CalculateEnergyFromPower(this,
+			ElectricityMeter.ChannelId.ACTIVE_CONSUMPTION_ENERGY);
+
+	private final Logger log = LoggerFactory.getLogger(IoShellyPlugsImpl.class);
+	private final BooleanWriteChannel[] digitalOutputChannels;
+
+	private MeterType meterType = null;
+	private SinglePhase phase = null;
+	private String baseUrl;
+
+	private volatile Timedata timedata;
+
+	@Reference(cardinality = ReferenceCardinality.MANDATORY)
+	private BridgeHttpFactory httpBridgeFactory;
+	private BridgeHttp httpBridge;
+
+	public IoShellyPlugsImpl() {
+		super(//
+				OpenemsComponent.ChannelId.values(), //
+				ElectricityMeter.ChannelId.values(), //
+				DigitalOutput.ChannelId.values(), //
+				IoShellyPlugs.ChannelId.values() //
+		);
+		this.digitalOutputChannels = new BooleanWriteChannel[] { //
+				this.channel(IoShellyPlugs.ChannelId.RELAY) //
+		};
+
+		SinglePhaseMeter.calculateSinglePhaseFromActivePower(this);
+	}
+
+	@Activate
+	protected void activate(ComponentContext context, Config config) {
+		super.activate(context, config.id(), config.alias(), config.enabled());
+		this.meterType = config.type();
+		this.phase = config.phase();
+		this.baseUrl = "http://" + config.ip();
+		this.httpBridge = this.httpBridgeFactory.get();
+
+		if (!this.isEnabled()) {
+			return;
+		}
+
+		this.httpBridge.subscribeJsonEveryCycle(this.baseUrl + "/rpc/Shelly.GetStatus", this::processHttpResult);
+	}
+
+	@Override
+	@Deactivate
+	protected void deactivate() {
+		if (this.httpBridge != null) {
+			this.httpBridgeFactory.unget(this.httpBridge);
+			this.httpBridge = null;
+		}
+		super.deactivate();
+	}
+
+	@Override
+	public BooleanWriteChannel[] digitalOutputChannels() {
+		return this.digitalOutputChannels;
+	}
+
+	@Override
+	public String debugLog() {
+		var b = new StringBuilder();
+		var valueOpt = this.getRelayChannel().value().asOptional();
+		if (valueOpt.isPresent()) {
+			b.append(valueOpt.get() ? "ON" : "OFF");
+		} else {
+			b.append("Unknown");
+		}
+		b.append("|");
+		b.append(this.getActivePowerChannel().value().asString());
+
+		return b.toString();
+	}
+
+	@Override
+	public void handleEvent(Event event) {
+		this.calculateEnergy();
```

Comment:
As you subscribe to two events, this method is called twice every Cycle - even if the Component is disabled.